### PR TITLE
Fixing operations on Mixed types

### LINKED ISSUE #7231: Table type checks now allow Mixed columns if all values satisfy a more precise type, but the Storage vectorized operations are not prepared for it
Since adding 'precise type detection' for Mixed columns in #6973, we are able to accept a Mixed type column wherever a column with more precise type is needed (e.g. a numeric column).

Unfortunately, our internal architecture is mostly not ready for this.

```python
from Standard.Base import all
from Standard.Table import all

main =
    t0 = Table.new [["X", ["a", 1, 2]], ["Y", [4, 5, 6]], ["Z", [7, 8, 9]]]
    t = t0.drop 1 . cast "Z" Value_Type.Mixed
    x = t.at "X"
    y = t.at "Y"
    z = t.at "Z"

    IO.println ("X: " + x.value_type.to_text + " / " + x.inferred_precise_value_type.to_text)
    IO.println ("Y: " + y.value_type.to_text)
    IO.println ("Z: " + z.value_type.to_text + " / " + z.inferred_precise_value_type.to_text)

    # This miraculously works, because MixedStorageFacade delegates to the underlying LongStorage - not sure if that is correct.
    IO.println <| Panic.recover Any (z+y)

    # This fails because ObjectStorage does not have vectorized `+`.
    IO.println <| Panic.recover Any (x+y)

    # This will fail with an internal type error - LongNumericOp is only taught to handle LongStorage and DoubleStorage.
    IO.println <| Panic.recover Any (y+x)
    IO.println <| Panic.recover Any (y+z)
```
prints
```
(In-Memory Column [Z] + [Y])
(Error: java.lang.IllegalArgumentException: The operation + has no vectorized implementation for this storage type, but no fallback function was provided. This is a bug in the Table library.)
(Error: org.enso.table.error.UnexpectedTypeException: Unexpected value type. Expected a Number..)
(Error: org.enso.table.error.UnexpectedTypeException: Unexpected value type. Expected a Number..)
```
exposing internal errors to the user.

In the past, all of these operations would fail with `Invalid_Value_Type`. The new approach is more lenient - we indeed want these mixed columns to 'pretend' they have this more specific type. To do so, we need to adapt our architecture accordingly.

So the expected behaviour is that all 4 of these operations should return a new numeric column, without errors.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Fixes #7231 
- Cleans up vectorized operations to distinguish unary and binary operations.
- Introduces MixedStorage which may pretend to be a more specialized storage on demand.
- Ensures that operations request a more specialized storage on right-hand side to ensure compatibility with reported inferred storage type.
- Ensures that a dataflow error returned by an Enso callback in Java is propagated as a polyglot exception and can be caught back in Enso
- Tests for comparison of Mixed storages with each other and other types
- Started using `Set` for `Filter_Condition.Is_In` for better performance.
- ~~Migrated `Column.map` and `Column.zip` to use the Java-to-Enso callbacks.~~
	- This does not forward warnings. IMO we should not be losing them. We can switch and add a ticket to fix the warnings, but that would be a regression (current implementation handles them correctly). Instead, we should first gain some ability to work with warnings in polyglot. I created a ticket to get this figured out #7371
- ~~Trying to avoid conversions when calling Enso functions from Java.~~
	- Needs extra care as dataflow errors may not be handled right then. So only works for simple functions that should not error.
	- Not sure how much it really helps. [Benchmarks](https://github.com/enso-org/enso/pull/7270#issuecomment-1635618393) suggested it could improve the performance quite significantly, but the practical solution is not exactly the same as the one measured, so we may have to measure and tune it to get the best results.
	- Created #7378 to track this.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso [hunk 1]
```diff
@@ -989,9 +1003,9 @@ type Column
         result = case self.value_type of
             Value_Type.Char _ _ -> self.is_empty
             Value_Type.Float _ ->
-                if treat_nans_as_blank then self.is_nothing.iif True (self.internal_is_nan on_missing=False) else self.is_nothing
+                if treat_nans_as_blank then self.is_nothing.iif True self.internal_is_nan else self.is_nothing
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso:1009]: Is this computing `self.is_nothing` twice?

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso:1009]: Good observation!! Yes, apparently it is.
Well, if it's a `LongStorage`, then it is a O(1) operation, because it just re-uses the bitset used to mark missing values. But I guess better to cache it just to be sure.

**Comment by radeusgd** [comment]
On occassion of limitting when the `convertPolyglotValue` helper is invoked (as the #7270 benchmarks showed that it accounts for a significant part of the runtime of various operations), I tried checking if it is even "still" needed.

After reminding myself of why it has been added in the first place, I conclude that it is needed and it will be needed whenever we are accepting date-time values, at least as long as we keep the architecture like the one we have currently.

1. The Graal Polyglot architecture does not auto-convert our Enso date-time values (`EnsoDate`, `EnsoDateTime`, `EnsoTimeOfDay`) into their Java counterparts, like it does for primitive types (converting Enso `Text` to Java `String`) and this is rather unlikely to change as it seems to be done _on purpose_.

The Graal polyglot values essentially can report `isDate`, `isTime` and `isZone`, specifying if they hold a date, a time-of-day and a timezone. A single value can hold any combination of these three. Graal cannot autoconvert our `EnsoDateTime` value (which reports `true` to all three) to `ZonedDateTime`. Technically, it probably could - but it does not do it. I hypothesize it may be because there may exsit different date-time implementations between languages, so there is no guarantee that such conversion would be lossless (in Enso its essentially a wrapper, so it would be, but that is not guaranteed in any language) and so it is not certain if it should be performed - so Graal leaves that decision to the _consumer_ of the value.

2. Our current architecture stores date-time values in in-memory Table Columns as Java values (`LocalDate`, `LocalTime` and `ZonedDateTime`). So whenever we get an `Object/Value` from Enso, we need to inspect if it should be coerced to the date, because as noted in (1) it will **not** be auto-converted - and that is what `convertPolyglotValue` is doing.

If we wanted to get rid of `convertPolyglotValue`, we'd need to somehow change the architecture. Probably to be more Graal-friendly and instead store date/time columns as `Object[]` or `Value[]` and then convert to dates more 'lazily'. To get the date/time information from the value in Java we'd have to do the kind of conversion that `convertPolyglotValue` does anyway at some point and it is unclear whether doing it more lazily would be any better.

Alternatively, **hypothetically**, to make it 'faster' we could try getting rid of `EnsoDate` et al. and using raw `LocalDate` and other Java values in Enso - then the conversion would be necessary. That is of course not a great idea, since the whole `EnsoDate` wrappers were done for a reason, to get a more user friendly and Enso-centric API for `Date`. It would also not fully solve the problem, because we can still get date-time values into our columns that come from other languages (e.g. JS, Python).

All-in-all I don't think we can avoid `convertPolyglotValue` in general (if I'm missing something here and it is possible, I'd be happy to be proved wrong 🙂). What we can do is try running it only if it is really needed - when we are building columns of number where any value that is not a number will be rejected anyway, calling the conversion will be unnecessary. So we only need to call it if the value may be a date and won't be rejected if it is a date.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso:1009]: No sorry, I'm reading the code again and realise it's not the case.

See how it's bracketed:
```
if treat_nans_as_blank 
	then self.is_nothing.iif True self.internal_is_nan 
	else self.is_nothing
```
So we can see that in each branch, `is_nothing` is computed once.

If only we had multiline if to make stuff like this more readable...

-----

Only thing we could optimize here is to compute `self.is_nothing.iif True self.internal_is_nan` in a single pass, by creating a variant of `internal_is_nan` that will return `True` for `null` values. But that seems out of scope of this PR.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Filter_Condition.enso [hunk 2]
```diff
@@ -192,10 +193,12 @@ type Filter_Condition
         Not_Like sql_pattern ->
             regex = sql_like_to_regex sql_pattern
             elem -> regex.matches elem . not
-        ## TODO once we have proper hashing we could create a hashmap and
-           answer quicker, currently we need to do a full scan for each element.
-        Is_In values -> values.contains
-        Not_In values -> elem -> values.contains elem . not
+        Is_In values ->
+            set = Set.from_vector values
+            set.contains
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Filter_Condition.enso:198]: shouldn't be the implementation of `contains` within `Vector`?

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/map/BinaryMapOperation.java [hunk 3]
```diff
@@ -27,7 +27,7 @@ public MapOperation(String name) {
    * @param problemBuilder the builder allowing to report computation problems
    * @return the result of running the operation
    */
-  public abstract Storage<?> runMap(
+  public abstract Storage<?> runBiMap(
```
Comment (jdunkerley) [inline std-bits/table/src/main/java/org/enso/table/data/column/operation/map/BinaryMapOperation.java]: Thought generally we avoided truncated names so should probably be `runBinaryMap`.

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/map/MapOpStorage.java [hunk 4]
```diff
@@ -11,40 +11,65 @@
...
+   * Runs the specified operation in map node.
    *
    * @param n the operation name
    * @param storage the storage to run operation on
    * @param arg the argument to pass to the operation
    * @param problemBuilder the builder allowing to report computation problems
    * @return the result of running the operation
    */
-  public Storage<?> runMap(
+  public Storage<?> runBiMap(
```
Comment (jdunkerley) [inline std-bits/table/src/main/java/org/enso/table/data/column/operation/map/MapOpStorage.java]: As above - `runBinaryMap`.

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/map/UnaryMapOperation.java [hunk 5]
```diff
@@ -3,24 +3,21 @@
...
+    this.name = name;
   }
 
-  protected abstract Storage<?> run(I storage);
-
-  @Override
-  public Storage<?> runMap(I storage, Object arg, MapOperationProblemBuilder problemBuilder) {
-    return run(storage);
-  }
+  protected abstract Storage<?> run(I storage, MapOperationProblemBuilder problemBuilder);
```
Comment (jdunkerley) [inline std-bits/table/src/main/java/org/enso/table/data/column/operation/map/UnaryMapOperation.java]: Lets either call this `runUnaryMap` to be consistent with `BinaryMapOperation` or better I think to rename `runBiMap` to just be `run` as well.

File: std-bits/table/src/main/java/org/enso/table/data/column/storage/Storage.java [hunk 6]
```diff
@@ -194,35 +195,24 @@ public final Storage<?> map(
...
-    if (name != null && isOpVectorized(name)) {
-      return runVectorizedZip(name, arg, problemBuilder);
-    }
-
-    checkFallback(function, expectedResultType, name);
-
+      StorageType expectedResultType) {
     Builder storageBuilder = Builder.getForType(expectedResultType, size());
-    Context context = Context.getCurrent();
+    final Context context = Context.getCurrent();
```
Comment (jdunkerley) [inline std-bits/table/src/main/java/org/enso/table/data/column/storage/Storage.java]: why this change?
Agree we don't change it but seems unneccessary.

File: test/Table_Tests/src/Common_Table_Operations/Main.enso [hunk 7]
```diff
@@ -97,7 +98,9 @@ type Test_Selection
...
          nanosecond precision in time values.
-    Config supports_case_sensitive_columns=True order_by=True natural_ordering=False case_insensitive_ordering=True order_by_unicode_normalization_by_default=False case_insensitive_ascii_only=False take_drop=True allows_mixed_type_comparisons=True supports_unicode_normalization=False is_nan_and_nothing_distinct=True distinct_returns_first_row_from_group_if_ordered=True date_time=True fixed_length_text_columns=False supports_decimal_type=False supports_time_duration=False supports_nanoseconds_in_time=False
+       - supports_mixed_columns: Specifies if the backend supports mixed-type
+         columns.
+    Config supports_case_sensitive_columns=True order_by=True natural_ordering=False case_insensitive_ordering=True order_by_unicode_normalization_by_default=False case_insensitive_ascii_only=False take_drop=True allows_mixed_type_comparisons=True supports_unicode_normalization=False is_nan_and_nothing_distinct=True distinct_returns_first_row_from_group_if_ordered=True date_time=True fixed_length_text_columns=False supports_decimal_type=False supports_time_duration=False supports_nanoseconds_in_time=False supports_mixed_columns=False
```
Comment (jdunkerley) [inline test/Table_Tests/src/Common_Table_Operations/Main.enso]: this is getting very long...
I suggest we drop `supports_` from names. Not necessarily in this PR!

**Review by jdunkerley — APPROVED** [review]
Looks good and great to have this more sorted.
One slight name sort out please.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Filter_Condition.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Filter_Condition.enso:198]: There is one, but using set will be faster.

This way here we construct the Set (O(N)) operation once, and save that set in the closure of the generated function, so then we can use the quick (~O(1), hashing-based) set checks.

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/map/BinaryMapOperation.java (hunk 3 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/data/column/operation/map/BinaryMapOperation.java]: Fair!

Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/data/column/operation/map/BinaryMapOperation.java]: I was keeping the original nomenclature (we used to have `bimap` here), but I think you are right, we should try to be consistent, I will rename.

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/map/UnaryMapOperation.java (hunk 5 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/data/column/operation/map/UnaryMapOperation.java]: But `runBiMap` has 2 variants - `runBiMap` and `runZip`, if we rename one of these to just `run` IMO it will be less clear what is the relation between `run` and `runZip`. So I'd go for `runUnaryMap` here.

File: std-bits/table/src/main/java/org/enso/table/data/column/storage/Storage.java (hunk 6 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/data/column/storage/Storage.java]: I was adding something else here and made this one final too. I can revert, not sure if this indeed changes anything - probably should not.

Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/data/column/storage/Storage.java]: Reverted
